pkg/gather/collect/collect.go

```diff
@@ -0,0 +1,489 @@
+	apierrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/dynamic"
+	corev1client "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/klog/v2"
+)
+
+const (
+	namespacesDirName    = "namespaces"
+	clusterScopedDirName = "cluster-scoped"
+)
+
+type ResourceInfo struct {
+	Scope    meta.RESTScope
+	Resource schema.GroupVersionResource
+}
+
+func NewResourceInfoFromMapping(mapping *meta.RESTMapping) *ResourceInfo {
+	return &ResourceInfo{
+		Scope:    mapping.Scope,
+		Resource: mapping.Resource,
+	}
+}
+
+func writeObject(printer ResourcePrinterInterface, filePath string, resourceInfo *ResourceInfo, obj kubeinterfaces.ObjectInterface) error {
+	buf := bytes.NewBuffer(nil)
+	err := printer.PrintObj(resourceInfo, obj, buf)
+	if err != nil {
+		return fmt.Errorf("can't print object %q (%s): %w", naming.ObjRef(obj), resourceInfo.Resource, err)
+	}
+
+	err = os.WriteFile(filePath, buf.Bytes(), 0770)
+	if err != nil {
+		return fmt.Errorf("can't write file %q: %w", filePath, err)
+	}
+
+	klog.V(4).InfoS("Written resource", "Path", filePath)
+
+	return nil
+}
+
+func getResourceKey(obj *unstructured.Unstructured, resourceInfo *ResourceInfo) string {
+	var nsPrefix string
+	if resourceInfo.Scope.Name() == meta.RESTScopeNameNamespace {
+		nsPrefix = obj.GetNamespace() + "/"
+	}
+
+	return nsPrefix + fmt.Sprintf("%s/%s/%s", resourceInfo.Resource.Group, resourceInfo.Resource.Resource, obj.GetName())
+}
+
+type Collector struct {
+	baseDir          string
+	printers         []ResourcePrinterInterface
+	discoveryClient  discovery.DiscoveryInterface
+	corev1Client     corev1client.CoreV1Interface
+	dynamicClient    dynamic.Interface
+	relatedResources bool
+	keepGoing        bool
+
+	collectedResources sets.Set[string]
+}
+
+func NewCollector(
+	baseDir string,
+	printers []ResourcePrinterInterface,
+	discoveryClient discovery.DiscoveryInterface,
+	corev1Client corev1client.CoreV1Interface,
+	dynamicClient dynamic.Interface,
+	relatedResources bool,
+	keepGoing bool,
+) *Collector {
+	return &Collector{
+		baseDir:            baseDir,
+		printers:           printers,
+		discoveryClient:    discoveryClient,
+		corev1Client:       corev1Client,
+		dynamicClient:      dynamicClient,
+		relatedResources:   relatedResources,
+		keepGoing:          keepGoing,
+		collectedResources: sets.Set[string]{},
+	}
+}
+
+func (c *Collector) getResourceDir(obj kubeinterfaces.ObjectInterface, resourceInfo *ResourceInfo) string {
+	scope := resourceInfo.Scope.Name()
+	switch scope {
+	case meta.RESTScopeNameNamespace:
+		return filepath.Join(
+			c.baseDir,
+			namespacesDirName,
+			obj.GetNamespace(),
+			resourceInfo.Resource.GroupResource().String(),
+		)
+
+	case meta.RESTScopeNameRoot:
+		return filepath.Join(
+			c.baseDir,
+			clusterScopedDirName,
+			resourceInfo.Resource.GroupResource().String(),
+		)
+
+	default:
+		return fmt.Sprintf("unrecognized scope %q", scope)
```

Comment:
shouldn't this return error in this case? Otherwise caller will try to write into `unrecognized [...]` directory

Reply:
yeah, this is weird